Week2/LESSONPLAN.md

```diff
@@ -0,0 +1,483 @@
+Demonstrate with four programs at [this repository](https://github.com/unmeshvrije/database_examples)
+how
+1. Program `1-db-naive.js` fails because the connection is closed prematurely.
+2. Program `2-db-callback.js` solves the problem but looks ugly because of the callback-hell.
+3. Program `3-db-promise.js` uses the promise chaining to make it better.
+about building those promises
+4. Program `4-db-await.js` uses promisify() and await() to make it the best.
+
+#### Exercise
+The program called `async-create-insert.js` can be found in `Week2` folder.
+Add a select query to that program using await and promisify.
+
+#### Essence
+> async keyword : to create asynchronous function and ensure they return promise without having to worry
+
+> await : to call a function returning promise without having to call .then() over that promise
+
+> promisify() : to convert a callback based function to a promise based one.
+
+
+### 1. Key (Constraints)
+#### Explanation
+1. A column can be declared as the UNIQUE column. Such a column has UNIQUE values.
+It can also have NULL values. Thus, two rows can have same NULL value in the column that
+is declared as UNIQUE (In other words, this is a UNIQUE CONSTRAINT on that column).
+2. A column can be declared as the PRIMARY KEY column. Such a column has UNIQUE values too.
+They cannot be NULL values. Thus two rows can NEVER have same values in the column
+that is declared as PRIMARY KEY (In other words, this is a PRIMARY KEY CONSTRAINT on that column).
+
+> There are more constraints in MySQL. Read more about them [here](https://www.w3resource.com//creating-table-advance/constraint.php).
+
+#### Example
+Consider the following commands (# represents comments).
+```sql
+# create table with two columns. one with primary key and one with unique key constraint
+CREATE TABLE pri_uniq_demo(id_pr int PRIMARY KEY, id_un int UNIQUE);
+
+# Note the error that says that the primary key column cannot be NULL
+INSERT INTO pri_uniq_demo VALUES (NULL, NULL);
+#ERROR 1048 (23000): Column 'id_pr' cannot be null
+
+# Note that the UNIQUE key column can be NULL
+INSERT INTO pri_uniq_demo VALUES (1, NULL);
+#Query OK, 1 row affected (0.00 sec)
+
+# Normal insertion
+INSERT INTO pri_uniq_demo VALUES (2, 2);
+#Query OK, 1 row affected (0.05 sec)
+
+# Note that you cannot insert 2 in the id_un column because it should be UNIQUE
+INSERT INTO pri_uniq_demo VALUES (3, 2);
+#ERROR 1062 (23000): Duplicate entry '2' for key 'id_un'
+
+# Note that you cannot insert 2 in the id_pr column because it is PRIMARY KEY
+INSERT INTO pri_uniq_demo VALUES (2, 3);
+#ERROR 1062 (23000): Duplicate entry '2' for key 'PRIMARY'
+
+```
+
+#### Exercise
+```
+# Find out type T and Constraint C for each column.
+CREATE TABLE Airline_passengers(ticket_numer T C, passenger_name T C, date_of_birth T C, passport_number T C);
+
+Hint: A very young baby may not need a ticket!
+```
+
+#### Essence
+Primary key is a special case of UNIQUE key. UNIQUE key can be NULL and
```

Comment:
Maybe a word on INDEX that comes with PK and UNIQUE could be useful. forgetting to add an index on a column that is searchable is still a really common bug.

Reply:
I tried very hard to not use the word INDEX because it can further create confusion with keys. But you do have a point. How can we describe it ? More importantly how could we demonstrate it in the class? I looked for some ways to populate database with lot of records and then see query performance difference with INDEXed columns and not INDEXed ones. But it is time consuming.

Reply:
I understand the confusion. I would explain the Index as a big huge MAP that gives the address of the row really quickly compare to no index and having to check each row by row.
I have just assigned you a PR with a quick run to see the difference between index and no index. Hope it will be helpful.